File: CPP/No523.cc
/**
 * Created by Xiaozhong on 2020/10/11.
 * Copyright (c) 2020/10/11 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool checkSubarraySum(vector<int> &nums, int k) {
        int sum = 0;
        unordered_map<int, int> mapper;
        mapper[0] = -1;

        for (int i = 0; i < nums.size(); ++i) {
            sum += nums[i];
            if (k != 0) sum %= k;
            if (mapper.count(sum)) {
                if (i - mapper[sum] > 1) return true;
            } else {
                mapper[sum] = i;
            }
        }
        return false;
    }
};

int main() {
    Solution s;
    vector<int> nums = {23, 2, 4, 6, 7};
    cout << s.checkSubarraySum(nums, -6) << endl;
    nums = {0, 0};
    cout << s.checkSubarraySum(nums, 0) << endl;
    nums = {0};
    cout << s.checkSubarraySum(nums, 0) << endl;
}

File: Java/src/No523.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.HashMap;
import java.util.Map;

public class No523 {
    public boolean checkSubarraySum(int[] nums, int k) {
        int n = nums.length;
        if (n < 2) return false;
        Map<Integer, Integer> mapper = new HashMap<>();
        mapper.put(0, -1);
        int remainder = 0;
        for (int i = 0; i < n; ++i) {
            remainder = (remainder + nums[i]) % k;
            if (mapper.containsKey(remainder)) {
                int prevIndex = mapper.get(remainder);
                if (i - prevIndex >= 2) return true;
            } else mapper.put(remainder, i);
        }
        return false;
    }
}
